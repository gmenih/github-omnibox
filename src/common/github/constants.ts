export const DEFAULT_FIRST_LOGINS: number = 100;
export const DEFAULT_FIRST_REPOS: number = 5;
export const DEFAULT_SCOPES: string[] = ['repo', 'read:org'];
export const GITHUB_API: string = 'https://api.github.com/graphql';
export const GITHUB_OAUTH_URL: string = 'https://github.com/login/oauth/authorize';
export const GITHUB_TOKEN_URL: string = 'https://github.com/login/oauth/access_token';
export const REDIRECT_URL: string = process.env.REDIRECT_URL || 'https://github.com/gmenih341/github-omnibox';
export const CLIENT_ID: string = process.env.CLIENT_ID || '';
export const CLIENT_SECRET: string = process.env.CLIENT_SECRET || '';
