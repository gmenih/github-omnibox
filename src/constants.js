export const GITHUB_API = 'https://api.github.com/graphql';
export const GITHUB_OAUTH_URL = 'https://github.com/login/oauth/authorize';
export const GITHUB_TOKEN_URL = 'https://github.com/login/oauth/access_token';
export const REDIRECT_URL = 'https://github.com/gmenih341/github-omnibox';
export const SEARCH_DEBOUNCE = 200;
export const OPTION_STRINGS = {
    GITHUB_TOKEN: '__github.token',
    GITHUB_LOGINS: '__github.logins',
    SEARCH_GLOBAL: '__github.search_global',
    SEARCH_DESC: '__github.search_desc',
    SEARCH_FORKED: '__github.search_forked',

    USE_HISTORY: '__extension.use_history',
    HISTORY: '__extension.search_history',
    UPDATE_LOGINS: '__extension.update_logins',
    OPTIONS_SHOWN: '__extension.options_shown_once',
    TOKEN_TYPE: '__extension.token_type',
};
