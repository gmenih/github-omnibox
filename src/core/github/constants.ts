export const DEFAULT_FIRST_LOGINS = 100;
export const DEFAULT_FIRST_REPOS = 5;
export const DEFAULT_SCOPES: string[] = ['repo:status', 'read:org'];
export const GITHUB_API = 'https://api.github.com/graphql';
export const GITHUB_OAUTH_URL = 'https://github.com/login/oauth/authorize';
export const GITHUB_TOKEN_URL = 'https://github.com/login/oauth/access_token';
export const REDIRECT_URL = 'https://github.com/gmenih341/github-omnibox';
export const CLIENT_ID = '';
export const CLIENT_SECRET = '';
