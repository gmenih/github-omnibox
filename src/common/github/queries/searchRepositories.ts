import {GraphQLClient} from 'graphql-request';
import {DEFAULT_FIRST_REPOS} from '../constants';

interface RepositoryNode {
    node: {
        name: string;
        url: string;
        owner: {
            login: string;
        };
    };
}

interface SearchResponse {
    search: {
        edges: RepositoryNode[];
    };
}

export interface GithubRepository {
    name: string;
    owner: string;
    url: string;
}

export async function searchRepositories (
    githubClient: GraphQLClient,
    searchTerm: string,
    pageSize: number = DEFAULT_FIRST_REPOS,
): Promise<GithubRepository[]> {
    try {
        const response: SearchResponse = await githubClient.request<SearchResponse>(
            /* GraphQL */ `
                query searchRepos($searchTerm: String!, $pageSize: Int!) {
                    search(query: $searchTerm, first: $pageSize, type: REPOSITORY) {
                        edges {
                            node {
                                ... on Repository {
                                    name
                                    url
                                    owner {
                                        login
                                    }
                                }
                            }
                        }
                    }
                }
            `,
            {
                pageSize,
                searchTerm,
            },
        );
        return response.search.edges.map(
            (edge): GithubRepository => ({
                name: edge.node.name,
                owner: edge.node.owner.login,
                url: edge.node.url,
            }),
        );
    } catch (err) {
        throw new Error('Failed to search repositories!');
    }
}
