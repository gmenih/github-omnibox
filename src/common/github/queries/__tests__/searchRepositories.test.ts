import {GraphQLClient} from 'graphql-request';
import {DEFAULT_FIRST_REPOS} from '../../constants';
import {searchRepositories} from '../searchRepositories';

const stubRepositoriesResponse = {
    search: {
        edges: [
            {
                node: {
                    name: 'test-repo',
                    owner: {
                        login: 'test-user',
                    },
                    url: 'https://github.com/test/repo',
                },
            },
        ],
    },
};

describe('Github searchRepositories', () => {
    let mockGraphQLClient: GraphQLClient;

    beforeEach(() => {
        mockGraphQLClient = {
            request: jest.fn(),
        } as any;
    });

    test('query matches snapshot and is called with proper arguments', async () => {
        let calledQuery: string = '';
        (mockGraphQLClient.request as jest.Mock).mockImplementation((query: string) => {
            calledQuery = query;
            return stubRepositoriesResponse;
        });

        await searchRepositories(mockGraphQLClient, 'test term');

        expect(mockGraphQLClient.request).toHaveBeenCalledTimes(1);
        expect(mockGraphQLClient.request).toHaveBeenCalledWith(expect.stringContaining('query'), {
            pageSize: DEFAULT_FIRST_REPOS,
            searchTerm: 'test term',
        });
        expect(calledQuery).toMatchSnapshot();
    });

    test('maps data to GithubRepository interface', async () => {
        (mockGraphQLClient.request as jest.Mock).mockResolvedValueOnce(stubRepositoriesResponse);

        const result = await searchRepositories(mockGraphQLClient, 'test term');

        expect(result).toMatchSnapshot();
    });
});
