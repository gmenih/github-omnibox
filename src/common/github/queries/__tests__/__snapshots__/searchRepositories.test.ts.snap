// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github searchRepositories maps data to GithubRepository interface 1`] = `
Array [
  Object {
    "name": "test-repo",
    "owner": "test-user",
    "url": "https://github.com/test/repo",
  },
]
`;

exports[`Github searchRepositories query matches snapshot and is called with proper arguments 1`] = `
"
                query searchRepos($searchTerm: String!, $pageSize: Int!) {
                    search(query: $searchTerm, first: $pageSize, type: REPOSITORY) {
                        edges {
                            node {
                                ... on Repository {
                                    name
                                    url
                                    owner {
                                        login
                                    }
                                }
                            }
                        }
                    }
                }
            "
`;
